pds_github_util/release/_python_version.py

```diff
@@ -0,0 +1,195 @@
+        if not setupFile: return None
+        expr = self.getRegexp()
+        with open(setupFile, 'r') as inp:
+            for line in inp:
+                match = expr.search(line)
+                if match: return match.group(1).strip()
+        return None
+
+
+class SetupConfigDetective(_SetupDetective):
+    '''Detective that parses the ``seutp.cfg`` file for a declarative version'''
+    def getFile(self):
+        return 'setup.cfg'
+    def getRegexp(self):
+        return re.compile(r'^version\s*=\s*([^#\s]+)')
+
+
+class SetupModuleDetective(_SetupDetective):
+    '''Detective that parses the ``setup.py`` module for a programmatic version'''
+    def getFile(self):
+        return 'setup.py'
+    def getRegexp(self):
+        return re.compile(r'version\s*=\s*[\'"]([^\'"]+)[\'"]')
+
+
+# Functions
+# ---------
+
+
+def registerDetective(detective: VersionDetective):
+    '''✍️ Register the given ``detective`` with the set of potential detetives to use to detect
+    version information in a Python source tree.
+    '''
+    _detectives.add(detective)
+
+
+def getVersion(workspace=None):
+    '''🕵️ Get the version of a Python package in the given ``workspace``, or in the directory
+    given by the ``GITHUB_WORKSPACE`` environment variable if it's set and non-empty,
+    or the current working directory. Try several strategies to determine the version and
+    use the one that makes the "most valid" version string, or raise a ``NoVersionDetectedError``
+    if none of them look copacetic.
+    '''
+    _logger.info('🤔 Python getVersion called with workspace %s', workspace)
+
+    # Figure out where to work
+    gh = os.getenv('GITHUB_WORKSPACE')
+    workspace = os.path.abspath(workspace if workspace else gh if gh else os.getcwd())
+    _logger.debug('👣 The computed path is %s', workspace)
+
+    # Try each detective
+    versions = set()
+    for detectiveClass in _detectives:
+        detective = detectiveClass(workspace)
+        version = detective.detect()
+        _logger.debug('🔍 Detected version using %s is %r', detectiveClass.__name__, version)
+        if version:
+            # Validate it
+            try:
+                versionObj = packaging.version.parse(version)
+                if not isinstance(versionObj, packaging.version.LegacyVersion):
+                    # A newer and therefore better version, so add it
+                    versions.add(version)
+            except packaging.version.InvalidVersion:
+                # Invalid, we won't add it
+                pass
+
+    # What we're left with are all valid so go with the shortest I guess; i.e., if one detective
+    # said ``1.2.3`` but another said ``1.2.3.post4`` we prefer ``1.2.3``.
+    if len(versions) == 0:
+        raise NoVersionDetectedError()
+    versions = list(versions)
+    versions.sort(key=len)
+    version = versions[0]
+    _logger.debug('🏁 High confidence version is %s', version)
+    return version
```

Comment:
Would there be any utility in converting this function to `getVersions` (plural) and returning all matches, then adding a `getVersion` that always returns the shortest?

Reply:
I doubt it.